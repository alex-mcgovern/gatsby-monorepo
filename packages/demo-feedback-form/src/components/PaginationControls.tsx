import React, { useCallback, useContext, useMemo } from "react";
import type { DropdownItem } from "@alexmcgovern/boondoggle.design";
import {
  Box,
  Button,
  Card,
  SelectSingle,
} from "@alexmcgovern/boondoggle.design";
import { FirebaseContext } from "@alexmcgovern/firebase";
import {
  faAngleLeft,
  faAngleRight,
  faRightToBracket,
} from "@fortawesome/free-solid-svg-icons";
import { Link } from "gatsby";
import type { UsePaginatedCommentsStateShape } from "../utils/usePaginatedComments";
import { FeedbackFormDialog } from "./FeedbackFormDialog";

const PER_PAGE_DROPDOWN_ITEMS: Array<DropdownItem> = [
  {
    label: "5 per page",
    value: "5",
  },
  {
    label: "10 per page",
    value: "10",
  },
  {
    label: "25 per page",
    value: "25",
  },
  {
    label: "50 per page",
    value: "50",
  },
];

export function PaginationControls({
  canLoadNext,
  canLoadPrevious,
  currentPage,
  firstItemIndex = 0,
  lastItemIndex = 0,
  loadNext,
  loadPrevious,
  setPerPage,
  totalNbComments,
  totalNbPages,
  ...rest
}: Partial<UsePaginatedCommentsStateShape>) {
  const { user } = useContext(FirebaseContext) || {};

  /**
   * State passed to `Link` to login page, to return user here
   * after sign in.
   */
  const linkState = useMemo(() => {
    return { returnTo: "/" };
  }, []);

  const updatePerPage = useCallback(
    (newValue: DropdownItem) => {
      if (setPerPage) {
        setPerPage(Number(newValue.value));
      }
    },
    [setPerPage]
  );

  return (
    <Box
      position="sticky"
      top="0"
      paddingTop="spacing1"
      background="neutral_background_raised"
      zIndex="1"
      {...rest}
    >
      <Card
        boxShadow="lg"
        display="flex"
        gap="spacing1"
        alignItems="center"
        flexWrap="wrap"
      >
        {/** --------------------------------------------
         * Login / leave feedback button
         * ----------------------------------------------- */}

        {user ? (
          <FeedbackFormDialog />
        ) : (
          <Button
            size="lg"
            as={Link}
            iconLeading={faRightToBracket}
            state={linkState}
            to="/login"
            width="max-content"
          >
            Log in to leave feedback
          </Button>
        )}

        {/** --------------------------------------------
         * Pagination state display
         * ----------------------------------------------- */}

        <Box marginLeft="auto">
          <div>
            Page <b>{currentPage ? currentPage + 1 : 1}</b> of{" "}
            <b>{totalNbPages}</b>
          </div>

          <div>
            <b> {firstItemIndex + 1}</b>â€”<b>{lastItemIndex + 1} </b> of{" "}
            <b>{totalNbComments}</b>
          </div>
        </Box>

        {/** --------------------------------------------
         * Per page
         * ----------------------------------------------- */}

        <SelectSingle
          size="lg"
          placeholder="5"
          items={PER_PAGE_DROPDOWN_ITEMS}
          onValueChange={updatePerPage}
        />

        {/** --------------------------------------------
         * Next/previous buttons
         * ----------------------------------------------- */}

        <Button
          size="lg"
          appearance="primary"
          disabled={!canLoadNext}
          iconLeading={faAngleLeft}
          onClick={loadNext}
        >
          Older
        </Button>

        <Button
          size="lg"
          appearance="primary"
          disabled={!canLoadPrevious}
          iconTrailing={faAngleRight}
          onClick={loadPrevious}
        >
          Newer
        </Button>
      </Card>
    </Box>
  );
}
