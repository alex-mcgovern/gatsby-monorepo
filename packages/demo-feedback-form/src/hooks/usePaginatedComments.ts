import { useCallback, useContext, useEffect, useMemo, useState } from "react";
import { FirebaseContext } from "@alexmcgovern/firebase";
import {
  collection,
  endBefore,
  getCountFromServer,
  getFirestore,
  limit,
  orderBy,
  query,
  startAfter,
} from "firebase/firestore";
import { useCollection } from "react-firebase-hooks/firestore";

interface UsePaginatedCommentsArgs {
  commentsPerPage: number;
}

export interface PaginationStateShape {
  current: number;
  previous: number;
}

export function usePaginatedComments({
  commentsPerPage,
}: UsePaginatedCommentsArgs) {
  const { firebaseApp } = useContext(FirebaseContext);

  const collectionRef = useMemo(() => {
    if (firebaseApp) {
      return collection(
        getFirestore(firebaseApp),
        "feedback",
        "data",
        "comments"
      );
    }
    return undefined;
  }, [firebaseApp]);

  /** -----------------------------------------------------------------------------
   * Query comments with react-firebase-hooks/firestore
   *
   * Note: We are using state to store query function, not setting the initial state with
   * result of a function. This is so we can update it later and re-render to trigger hook.
   * ------------------------------------------------------------------------------- */

  const [queryFunction, setCachedQuery] = useState(() => {
    return () => {
      if (collectionRef) {
        return query(
          collectionRef,
          orderBy("created", "desc"),
          limit(commentsPerPage)
        );
      }
      return undefined;
    };
  });

  const [commentsValues, commentsLoading, commentsError] = useCollection(
    queryFunction(),
    {
      snapshotListenOptions: { includeMetadataChanges: true },
    }
  );

  const lastVisibleComment =
    commentsValues?.docs[commentsValues.docs.length - 1];

  const firstVisibleComment = commentsValues?.docs[0];

  const transformedComments = commentsValues?.docs.map((doc) => {
    return { ...doc.data(), documentRef: doc.ref };
  });

  /** -----------------------------------------------------------------------------
   * Manually track pagination state, attempt to sync with firestore state
   * We need this to disable buttons and prevent trying to access out of range
   * documents. We'll increment/decrement these via the pagination buttons.
   *
   * This isn't ideal, as this could potentially fall out of sync, but I wasn't
   * able to find a better & simpler alternative.
   *
   * Infinite scrolling might be an option, but I'm not a fan of that  pattern as
   * it can cause performance issues when loading in large amounts of data, and the
   * UX can be bad without being handled correctly (e.g. scroll to top handling).
   *
   * Windowing might be a better design choice, but probably a bit too complex for this small project.
   * ------------------------------------------------------------------------------- */

  const [commentsCount, setCommentsCount] = useState<number>(0);
  const [paginationState, setPaginationState] = useState<PaginationStateShape>({
    current: 0,
    previous: 0,
  });

  useEffect(() => {
    if (collectionRef) {
      getCountFromServer(collectionRef).then((response) => {
        return setCommentsCount(response.data().count);
      });
    }
  }, [collectionRef]);

  const { totalPages, canLoadOlderComments, canLoadNewerComments } =
    useMemo(() => {
      const totalNbPages =
        commentsCount && Math.floor(commentsCount / commentsPerPage);

      return {
        totalPages: totalNbPages,
        canLoadOlderComments:
          lastVisibleComment && paginationState.current < totalNbPages,
        canLoadNewerComments:
          firstVisibleComment && paginationState.current > 0,
      };
    }, [
      commentsCount,
      commentsPerPage,
      paginationState,
      firstVisibleComment,
      lastVisibleComment,
    ]);

  /** -----------------------------------------------------------------------------
   * Handle increment / decrement
   *
   * For each, create a new query function and trigger re-render by updating state.
   * Do checking to make sure we are not trying to access documents that are out of range.
   * ------------------------------------------------------------------------------- */

  const loadOlderComments = useCallback(() => {
    if (canLoadOlderComments) {
      setPaginationState((paginationStateState) => {
        return {
          current: paginationStateState.current + 1,
          previous: paginationStateState.current,
        };
      });

      setCachedQuery(() => {
        return () => {
          if (collectionRef) {
            return query(
              collectionRef,
              orderBy("created", "desc"),
              startAfter(lastVisibleComment),
              limit(commentsPerPage)
            );
          }
          return undefined;
        };
      });
    }
  }, [
    canLoadOlderComments,
    collectionRef,
    commentsPerPage,
    lastVisibleComment,
  ]);

  const loadNewerComments = useCallback(() => {
    if (canLoadNewerComments) {
      setPaginationState((paginationStateState) => {
        return {
          current: paginationStateState.current - 1,
          previous: paginationStateState.current,
        };
      });

      setCachedQuery(() => {
        return () => {
          if (collectionRef) {
            return query(
              collectionRef,
              orderBy("created", "desc"),
              endBefore(firstVisibleComment),
              limit(commentsPerPage)
            );
          }
          return undefined;
        };
      });
    }
  }, [
    canLoadNewerComments,
    collectionRef,
    commentsPerPage,
    firstVisibleComment,
  ]);

  /** -----------------------------------------------------------------------------
   * Return values
   * ------------------------------------------------------------------------------- */

  return {
    canLoadNewerComments,
    canLoadOlderComments,
    comments: transformedComments,
    commentsCount,
    commentsError,
    commentsLoading,
    paginationState,
    loadNewerComments,
    loadOlderComments,
    totalPages,
  };
}
