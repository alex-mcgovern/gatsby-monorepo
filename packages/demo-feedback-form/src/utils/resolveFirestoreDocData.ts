import type { DocumentData, QueryDocumentSnapshot } from "firebase/firestore";

export function resolveFirestoreDocData(
  documentSnapshots?: Array<QueryDocumentSnapshot<DocumentData>>
) {
  return documentSnapshots?.map((doc) => {
    return { ...doc.data(), documentRef: doc.ref };
  });
}
