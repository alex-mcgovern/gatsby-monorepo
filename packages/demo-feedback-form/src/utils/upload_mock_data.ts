import { initializeApp } from "firebase/app";
import {
  Timestamp,
  addDoc,
  collection,
  getFirestore,
} from "firebase/firestore";
import { COMMENTS_MOCK_DATA_SET } from "../../__mocks__/COMMENTS_MOCK_DATASET";

const firebaseConfig = {
  apiKey: process.env.FIREBASE_WEB_API_KEY,
  authDomain: process.env.FIREBASE_AUTH_DOMAIN,
  databaseURL: process.env.FIREBASE_DATABASE_URL,
  projectId: process.env.FIREBASE_PROJECT_ID,
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.FIREBASE_APP_ID,
  measurementId: process.env.FIREBASE_MEASUREMENT_ID,
};

const firebaseApp = initializeApp(firebaseConfig);

COMMENTS_MOCK_DATA_SET.forEach(
  async ({ displayName, email, rating, description, author_uid, created }) => {
    return addDoc(
      collection(getFirestore(firebaseApp), "feedback", "data", "comments"),
      {
        displayName,
        email,
        rating,
        description,
        author_uid,
        created: Timestamp.fromMillis(created),
      }
    ).catch((error) => {
      console.error(error);
    });
  }
);
