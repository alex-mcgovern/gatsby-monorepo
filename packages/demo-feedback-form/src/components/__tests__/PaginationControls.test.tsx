/** @jest-environment jsdom */
import React from "react";
import { cleanup } from "@testing-library/react";
import { renderTestComponent } from "../../../../../test/renderTestComponent";
import { UsePaginatedCommentsStateShape } from "../../utils/usePaginatedComments";
import { PaginationControls } from "../PaginationControls";

afterEach(cleanup);

/** ---------------------------------------------
 * Mock firestore when not directly testing as jest
 * will attempt to pull in un-transpiled es6 js from firebase
 * ----------------------------------------------- */
jest.mock("firebase/firestore", () => {
  return {
    getFirestore: jest.fn(),
  };
});
jest.mock("firebase/app", () => {
  return {
    initializeApp: jest.fn(),
  };
});
jest.mock("firebase/auth", () => {
  return {
    getAuth: jest.fn(),
  };
});

test("Given valid props, renders without error", () => {
  const { getByTestId } = renderTestComponent<UsePaginatedCommentsStateShape>(
    <PaginationControls
      canLoadNext={true}
      canLoadPrevious={false}
      currentPage={0}
      firstItemIndex={0}
      lastItemIndex={4}
      totalNbComments={101}
      totalNbPages={21}
    />
  );

  expect(getByTestId("tested-component")).not.toBeNull();
});

test("Given valid props, matches snapshot", () => {
  const { getByTestId } = renderTestComponent<UsePaginatedCommentsStateShape>(
    <PaginationControls
      canLoadNext={true}
      canLoadPrevious={false}
      currentPage={0}
      firstItemIndex={0}
      lastItemIndex={4}
      totalNbComments={101}
      totalNbPages={21}
    />
  );

  expect(getByTestId("tested-component")).toMatchSnapshot();
});

test("Given initial props, buttons have correct `disabled` state", () => {
  const { getByText } = renderTestComponent<UsePaginatedCommentsStateShape>(
    <PaginationControls
      canLoadNext={true}
      canLoadPrevious={false}
      currentPage={0}
      firstItemIndex={0}
      lastItemIndex={4}
      totalNbComments={101}
      totalNbPages={21}
    />
  );

  expect(getByText("Older")).not.toBeDisabled();
  expect(getByText("Newer")).toBeDisabled();
});

test("Given props for second page, buttons have correct `disabled` state", () => {
  const { getByText } = renderTestComponent<UsePaginatedCommentsStateShape>(
    <PaginationControls
      canLoadNext={true}
      canLoadPrevious={true}
      currentPage={1}
      firstItemIndex={10}
      lastItemIndex={19}
      totalNbComments={101}
      totalNbPages={21}
    />
  );

  expect(getByText("Older")).not.toBeDisabled();
  expect(getByText("Newer")).not.toBeDisabled();
});
