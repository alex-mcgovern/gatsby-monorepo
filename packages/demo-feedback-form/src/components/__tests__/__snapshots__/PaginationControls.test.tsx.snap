// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given valid props, matches snapshot 1`] = `
<div
  class="getSprinkles_position_sticky__1jganc9ai getSprinkles_top_0__1jganc9af getSprinkles_paddingTop_spacing1__1jganc9aa getSprinkles_background_neutral_background_raised__1jganc9q getSprinkles_zIndex_1__1jganc9at"
  data-testid="tested-component"
>
  <div
    class="Card_cardStyle__1s7cia51 Card__1s7cia50 getSprinkles_padding_spacing2__1jganc94v getSprinkles_display_block_mobile__1jganc9au getSprinkles_isolation_isolate__1jganc93j getSprinkles_textDecoration_none__1jganc9an getSprinkles_background_neutral_secondary_base__1jganc9y getSprinkles_border_neutral_border_interactive__1jganc92g getSprinkles_borderRadius_md__1jganc92a focus_ring_styles_focusedStateStyle__1vv8g8r0 getSprinkles_boxShadow_lg__1jganc98 getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_gap_spacing2__1jganc93v getSprinkles_alignItems_center__1jganc92 getSprinkles_flexWrap_wrap__1jganc93a"
  >
    <a
      class="button.theme_buttonTheme__1a9lj020 Button_getButtonStyles__1lclbjh1 focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 Button__1lclbjh0 getSprinkles_borderRadius_md__1jganc92a getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_gap_spacing0__1jganc93t getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an Button_getButtonStyles_appearance_primary__1lclbjh2 variantButtonAppearance_variantButtonAppearance_primary__gh61hw4 variantButtonAppearance__gh61hw0 getSprinkles_paddingLeft_spacing2__1jganc99h getSprinkles_paddingRight_spacing2__1jganc99w getSprinkles_justifyContent_center__1jganc941 getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_fontWeight_semibold__1jganc93g Button_getButtonStyles_size_md__1lclbjha variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq getSprinkles_width_max-content_mobile__1jganc9fo"
      href="/login"
      type="button"
    >
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-right-to-bracket icon_icon__111gfxw0 Button__1lclbjhf getSprinkles_flexShrink_0__1jganc93c"
          data-icon="right-to-bracket"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M352 96h64c17.7 0 32 14.3 32 32V384c0 17.7-14.3 32-32 32H352c-17.7 0-32 14.3-32 32s14.3 32 32 32h64c53 0 96-43 96-96V128c0-53-43-96-96-96H352c-17.7 0-32 14.3-32 32s14.3 32 32 32zm-7.5 177.4c4.8-4.5 7.5-10.8 7.5-17.4s-2.7-12.9-7.5-17.4l-144-136c-7-6.6-17.2-8.4-26-4.6s-14.5 12.5-14.5 22v72H32c-17.7 0-32 14.3-32 32v64c0 17.7 14.3 32 32 32H160v72c0 9.6 5.7 18.2 14.5 22s19 2 26-4.6l144-136z"
            fill="currentColor"
          />
        </svg>
      </div>
      Log in to leave feedback
    </a>
    <div
      class="getSprinkles_marginLeft_auto__1jganc95l"
    >
      <div>
        Page 
        <b>
          1
        </b>
         of
         
        <b>
          21
        </b>
      </div>
      <div>
        <b>
           
          1
        </b>
        â€”
        <b>
          5
           
        </b>
         of
         
        <b>
          101
        </b>
      </div>
    </div>
    <div
      class="getSprinkles_position_relative__1jganc9ag"
    >
      <button
        aria-activedescendant=""
        aria-controls="downshift-1-menu"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-1-label downshift-1-toggle-button"
        class="button.theme_buttonTheme__1a9lj020 Button_getButtonStyles__1lclbjh1 focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 Button__1lclbjh0 getSprinkles_borderRadius_md__1jganc92a getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_gap_spacing0__1jganc93t getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an Button_getButtonStyles_appearance_select__1lclbjh4 variantButtonAppearance_variantButtonAppearance_select__gh61hw6 variantButtonAppearance__gh61hw2 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_alignItems_center__1jganc92 getSprinkles_background_neutral_background_base__1jganc9p getSprinkles_border_neutral_border_interactive__1jganc92g getSprinkles_borderRadius_md__1jganc92a getSprinkles_color_neutral_text_lowContrast__1jganc91u getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_overflow_hidden__1jganc98p getSprinkles_width_100%_mobile__1jganc9f9 Button_getButtonStyles_size_md__1lclbjha variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_justifyContent_space-between__1jganc944"
        role="combobox"
        tabindex="0"
        type="button"
      >
        5
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-down icon_icon__111gfxw0 Button__1lclbjhf getSprinkles_flexShrink_0__1jganc93c"
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
      <div
        aria-labelledby="downshift-1-label"
        class="DropdownMenu_dropdownMenu__1ubprtm1 focus_ring_styles_focusedStateStyle__1vv8g8r0 DropdownMenu__1ubprtm0 getSprinkles_marginTop_spacing1__1jganc96i getSprinkles_width_max-content_mobile__1jganc9fo getSprinkles_minWidth_100%__1jganc97m getSprinkles_maxWidth_gridSpan3__1jganc97b getSprinkles_display_block_mobile__1jganc9au getSprinkles_margin_none__1jganc94d getSprinkles_borderRadius_sm__1jganc92b getSprinkles_overflow_hidden__1jganc98p getSprinkles_background_neutral_background_raised__1jganc9q getSprinkles_boxShadow_sm__1jganc9a getSprinkles_border_neutral_border_interactive__1jganc92g getSprinkles_overflowY_auto__1jganc98q DropdownMenu_dropdownMenuClosed__1ubprtm4"
        id="downshift-1-menu"
        role="listbox"
        tabindex="-1"
      >
        <div
          class="DropdownMenu_dropdownListInner__1ubprtm3 DropdownMenu__1ubprtm2"
        >
          <span
            aria-selected="false"
            class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_textAlign_left__1jganc9ak getSprinkles_color_neutral_text_highContrast__1jganc91v getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_margin_none__1jganc94d getSprinkles_gap_spacing0__1jganc93t focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
            id="downshift-1-item-0"
            role="option"
          >
            5 per page
          </span>
          <span
            aria-selected="false"
            class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_textAlign_left__1jganc9ak getSprinkles_color_neutral_text_highContrast__1jganc91v getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_margin_none__1jganc94d getSprinkles_gap_spacing0__1jganc93t focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
            id="downshift-1-item-1"
            role="option"
          >
            10 per page
          </span>
          <span
            aria-selected="false"
            class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_textAlign_left__1jganc9ak getSprinkles_color_neutral_text_highContrast__1jganc91v getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_margin_none__1jganc94d getSprinkles_gap_spacing0__1jganc93t focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
            id="downshift-1-item-2"
            role="option"
          >
            25 per page
          </span>
          <span
            aria-selected="false"
            class="DropdownListItem_getDropdownListItemStyles__mbnr2x1 DropdownListItem__mbnr2x0 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_textAlign_left__1jganc9ak getSprinkles_color_neutral_text_highContrast__1jganc91v getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_margin_none__1jganc94d getSprinkles_gap_spacing0__1jganc93t focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 DropdownListItem_getDropdownListItemStyles_size_md__mbnr2x3 variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
            id="downshift-1-item-3"
            role="option"
          >
            50 per page
          </span>
        </div>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="button.theme_buttonTheme__1a9lj020 Button_getButtonStyles__1lclbjh1 focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 Button__1lclbjh0 getSprinkles_borderRadius_md__1jganc92a getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_gap_spacing0__1jganc93t getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an Button_getButtonStyles_appearance_primary__1lclbjh2 variantButtonAppearance_variantButtonAppearance_primary__gh61hw4 variantButtonAppearance__gh61hw0 getSprinkles_paddingLeft_spacing2__1jganc99h getSprinkles_paddingRight_spacing2__1jganc99w getSprinkles_justifyContent_center__1jganc941 getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_fontWeight_semibold__1jganc93g Button_getButtonStyles_size_md__1lclbjha variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
      type="button"
    >
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-left icon_icon__111gfxw0 Button__1lclbjhf getSprinkles_flexShrink_0__1jganc93c"
          data-icon="angle-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
            fill="currentColor"
          />
        </svg>
      </div>
      Older
    </button>
    <button
      aria-disabled="true"
      class="button.theme_buttonTheme__1a9lj020 Button_getButtonStyles__1lclbjh1 focus_ring_styles_focusedStateStyle__1vv8g8r0 disabled_state_styles_disabledStateStyles__5jo1bk0 Button__1lclbjh0 getSprinkles_borderRadius_md__1jganc92a getSprinkles_display_flex_mobile__1jganc9ax getSprinkles_gap_spacing0__1jganc93t getSprinkles_alignItems_center__1jganc92 getSprinkles_textDecoration_none__1jganc9an Button_getButtonStyles_appearance_primary__1lclbjh2 variantButtonAppearance_variantButtonAppearance_primary__gh61hw4 variantButtonAppearance__gh61hw0 getSprinkles_paddingLeft_spacing2__1jganc99h getSprinkles_paddingRight_spacing2__1jganc99w getSprinkles_justifyContent_center__1jganc941 getSprinkles_whiteSpace_nowrap__1jganc9ar getSprinkles_fontWeight_semibold__1jganc93g Button_getButtonStyles_size_md__1lclbjha variant_interactive_element_size_variantInteractiveElementSize_md__t3r2od7 variant_interactive_element_size__t3r2od1 getSprinkles_fontSize_body_lg_mobile__1jganc9c3 getSprinkles_fontSize_body_md_desktop__1jganc9c8 getSprinkles_lineHeight_body_lg_mobile__1jganc9cx getSprinkles_lineHeight_body_md_desktop__1jganc9d2 getSprinkles_height_spacing5_mobile__1jganc9ir getSprinkles_height_spacing4_desktop__1jganc9iq"
      disabled=""
      type="button"
    >
      Newer
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right icon_icon__111gfxw0 Button__1lclbjhf getSprinkles_flexShrink_0__1jganc93c"
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
